fold/fold.go

```diff
@@ -0,0 +1,923 @@
+// i.e. stacks, bulges, hairpins, etc.
+func Fold(seq string, temp float64) ([]NucleicAcidStructure, error) {
+	foldContext, err := newFoldingContext(seq, temp)
+	if err != nil {
+		return nil, fmt.Errorf("error creating folding context: %w", err)
+	}
+
+	// get the minimum free energy structure out of the cache
+	return traceback(0, len(seq)-1, foldContext), nil
+}
+
+// MinimumFreeEnergy folds the sequence and return just the delta G of the structure
+// Args:
+//
+//	seq: The sequence to fold
+//	temp: The temperature to fold at
+//
+// Returns: the minimum free energy of the folded sequence
+func MinimumFreeEnergy(seq string, temp float64) (float64, error) {
+	NucleicAcidStructures, err := Fold(seq, temp)
+	if err != nil {
+		return 0, fmt.Errorf("error folding: %w", err)
+	}
+	summedEnergy := 0.0
+	for _, structure := range NucleicAcidStructures {
+		summedEnergy += structure.Energy
+	}
+	return summedEnergy, nil
+}
+
+// DotBracket returns the dot-bracket notation of a secondary nucleic acid structure
+// Args:
+//
+//	NucleicAcidStructures: A list of NucleicAcidStructure, usually from the fold function
+//
+// Returns the dot-bracket notation of the secondary structure.
+//
+// Dot-bracket notation, consisting in a balanced parentheses string composed
+// by a three-character alphabet {.,(,)}, that can be unambiguously converted
+// in the RNA secondary structure. See example_test.go for a small example.
+func DotBracket(NucleicAcidStructures []NucleicAcidStructure) string {
+	lastStructEnd := 0
+	for _, structure := range NucleicAcidStructures {
+		for _, innerSubsequence := range structure.Inner {
+			if innerSubsequence.End > lastStructEnd {
+				lastStructEnd = innerSubsequence.End
+			}
+		}
+	}
+	lastStructEnd += 1
+	result := make([]byte, lastStructEnd)
+	for i := range result {
+		result[i] = '.'
+	}
+	for _, structure := range NucleicAcidStructures {
+		if len(structure.Inner) == 1 {
+			innerSubsequence := structure.Inner[0]
+			result[innerSubsequence.Start] = '('
+			result[innerSubsequence.End] = ')'
+		}
+	}
+	return string(result)
+}
+
+// unpairedMinimumFreeEnergyW returns the minimum free energy of a subsequence
+// at start and terminating at end.
+//
+// Figure 2B in Zuker and Stiegler, 1981
+// Args:
+//
+//		seq: The sequence being folded
+//		start: The start index
+//		end: The end index (inclusive)
+//	 foldContext: The context for this sequence
+//
+// Returns the free energy for the subsequence from start to end
+func unpairedMinimumFreeEnergyW(start, end int, foldContext context) (NucleicAcidStructure, error) {
+	if !foldContext.unpairedMinimumFreeEnergyW[start][end].Equal(defaultStructure) {
+		return foldContext.unpairedMinimumFreeEnergyW[start][end], nil
+	}
+
+	if end-start < 4 {
+		foldContext.unpairedMinimumFreeEnergyW[start][end] = invalidStructure
+		return foldContext.unpairedMinimumFreeEnergyW[start][end], nil
+	}
+
+	w1, err := unpairedMinimumFreeEnergyW(start+1, end, foldContext)
```

Comment:
what are these different tiers of unpaired minimum free energies? what can we call them besides w1, w2, w3?

Reply:
yes found some better names from the paper